proxy.go

```diff
@@ -0,0 +1,333 @@
+package veneur
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/pprof"
+	"os"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/Sirupsen/logrus"
+	raven "github.com/getsentry/raven-go"
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/profile"
+	"github.com/stripe/veneur/discovery"
+	"github.com/stripe/veneur/samplers"
+	"github.com/stripe/veneur/trace"
+	"github.com/zenazn/goji/bind"
+	"github.com/zenazn/goji/graceful"
+	"stathat.com/c/consistent"
+
+	"goji.io"
+	"goji.io/pat"
+)
+
+type Proxy struct {
+	statsd *statsd.Client
+	sentry *raven.Client
+
+	Hostname               string
+	ForwardDestinations    *consistent.Consistent
+	TraceDestinations      *consistent.Consistent
+	Discoverer             discovery.Discoverer
+	ConsulForwardService   string
+	ConsulTraceService     string
+	ConsulInterval         time.Duration
+	ForwardDestinationsMtx sync.Mutex
+	TraceDestinationsMtx   sync.Mutex
+	HTTPAddr               string
+	HTTPClient             *http.Client
+
+	enableProfiling bool
+}
+
+func NewProxyFromConfig(conf ProxyConfig) (ret Proxy, err error) {
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		log.WithError(err).Error("Error finding hostname")
+		return
+	}
+	ret.Hostname = hostname
+
+	log.Hooks.Add(sentryHook{
+		c:        ret.sentry,
+		hostname: hostname,
+		lv: []logrus.Level{
+			logrus.ErrorLevel,
+			logrus.FatalLevel,
+			logrus.PanicLevel,
+		},
+	})
+
+	ret.HTTPAddr = conf.HTTPAddress
+	// TODO Timeout?
+	ret.HTTPClient = &http.Client{}
+
+	ret.enableProfiling = conf.EnableProfiling
+
+	ret.ConsulForwardService = conf.ConsulForwardServiceName
+	ret.ConsulTraceService = conf.ConsulTraceServiceName
+
+	ret.ForwardDestinations = consistent.New()
+	ret.TraceDestinations = consistent.New()
+
+	ret.ConsulInterval, err = time.ParseDuration(conf.ConsulRefreshInterval)
+	if err != nil {
+		log.WithError(err).Error("Error parsing Consul refresh interval")
+		return
+	}
+	log.WithField("interval", conf.ConsulRefreshInterval).Info("Will use Consul for service discovery")
+
+	// TODO Size of replicas in config?
+	//ret.ForwardDestinations.NumberOfReplicas = ???
+
+	return
+}
+
+func (p Proxy) GetHostname() string {
```

Comment:
I'd prefer to omit the `Get` in getters unless we need them for a reason (such as disambiguation with named fields).

e.g., `Hostname()`, `Sentry()`, `Stats()`, etc.